js_modules/dagit/packages/core/src/ui/Filters/TimeRangeFilter.tsx

```diff
@@ -0,0 +1,291 @@
+  label: TimeRanges[key].label,
+  range: TimeRanges[key].range,
+}));
+
+export class TimeRangeFilter extends Filter<TimeRangeState, TimeRangeKey> {
+  constructor(name: string, icon: IconName, initialState?: TimeRangeState) {
+    super(name, icon, initialState || [null, null]);
+  }
+
+  renderActiveFilterState(): JSX.Element | null {
+    return (
+      <ActiveFilterState
+        state={this.getState()}
+        remove={() => {
+          this.setState([null, null]);
+        }}
+      />
+    );
+  }
+
+  isActive(): boolean {
+    const [start, end] = this.getState();
+    return start !== null || end !== null;
+  }
+
+  getResults(query: string): {label: JSX.Element; value: TimeRangeKey}[] {
+    if (query === '') {
+      return timeRangesArray.map(({label, key}) => ({
+        label: <TimeRangeResult range={label} />,
+        value: key,
+      }));
+    }
+    return timeRangesArray
+      .filter(({label}) => label.toLowerCase().includes(query.toLowerCase()))
+      .map(({label, key}) => ({
+        label: <TimeRangeResult range={label} />,
+        value: key,
+      }));
+  }
+
+  onSelect(
+    key: TimeRangeKey,
+    close: () => void,
+    createPortal: (element: JSX.Element) => () => void,
+  ) {
+    if (key === 'CUSTOM') {
+      const closeRef = {
+        current: () => {
+          debugger;
+        },
+      };
+      closeRef.current = createPortal(
+        <CustomTimeRangeFilterDialog filter={this} closeRef={closeRef} />,
+      );
+    } else {
+      TimeRanges = calculateTimeRanges();
+      const value = TimeRanges[key].range;
+      this.setState(value);
+    }
+    close();
+    return null;
+  }
+}
+
+function TimeRangeResult({range}: {range: string}) {
+  return (
+    <Box flex={{direction: 'row', gap: 12, alignItems: 'center'}}>
+      <Icon name="date" color={Colors.Dark} />
+      {range}
+    </Box>
+  );
+}
+
+const L_FORMAT = new Intl.DateTimeFormat(navigator.language, {
+  year: 'numeric',
+  month: 'numeric',
+  day: 'numeric',
+  timeZone: 'UTC',
+});
+
+function ActiveFilterState({state, remove}: {state: TimeRangeState; remove: () => void}) {
+  const dateLabel = React.useMemo(() => {
+    if (isEqual(state, TimeRanges.TODAY.range)) {
+      return (
+        <>
+          is <FilterTagHighlightedText>Today</FilterTagHighlightedText>
+        </>
+      );
+    } else if (isEqual(state, TimeRanges.YESTERDAY.range)) {
+      return (
+        <>
+          is <FilterTagHighlightedText>Yesterday</FilterTagHighlightedText>
+        </>
+      );
+    } else if (isEqual(state, TimeRanges.LAST_7_DAYS.range)) {
+      return (
+        <>
+          is within <FilterTagHighlightedText>Last 7 days</FilterTagHighlightedText>
+        </>
+      );
+    } else if (isEqual(state, TimeRanges.LAST_30_DAYS.range)) {
+      return (
+        <>
+          is within <FilterTagHighlightedText>Last 30 days</FilterTagHighlightedText>
+        </>
+      );
+    } else {
+      if (!state[0]) {
```

Comment:
Worth using keys instead of a tuple for tracking ranges? I think it's not too bad as is, but it might be a little easier to read.